src/axom/core/tests/core_array.cpp

```diff
@@ -0,0 +1,1103 @@
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i);
+  }
+}
+
+template <typename T>
+void check_swap(Array<T>& v)
+{
+  axom::Array<T> v_two(v.size());
+
+  /* Push 0...size elements */
+  for(int i = 0; i < v.size(); i++)
+  {
+    v[i] = i;
+    v_two[i] = -i;
+  }
+
+  /* Create copies */
+  axom::Array<T> v_copy(v);
+  axom::Array<T> v_two_copy(v_two);
+
+  EXPECT_EQ(v, v_copy);
+  EXPECT_EQ(v_two, v_two_copy);
+
+  /* Swap */
+  v.swap(v_two);
+
+  EXPECT_EQ(v_two, v_copy);
+  EXPECT_EQ(v, v_two_copy);
+
+  /* Swap back */
+  v.swap(v_two);
+
+  EXPECT_EQ(v, v_copy);
+  EXPECT_EQ(v_two, v_two_copy);
+}
+
+template <typename T>
+void check_alloc(Array<T>& v, const int& id)
+{
+  // Verify allocation
+  EXPECT_EQ(v.getAllocatorID(), id);
+  EXPECT_EQ(v.size(), v.capacity());
+
+// Use introspection to verify pointer if available
+#if defined(AXOM_USE_UMPIRE)
+  umpire::ResourceManager& rm = umpire::ResourceManager::getInstance();
+  auto v_allocator = rm.getAllocator(v.data());
+  EXPECT_EQ(v_allocator.getId(), id);
+#endif
+}
+
+/*!
+ * \brief Check an external array for defects.
+ * \param [in] v the external array to check.
+ */
+template <typename T>
+void check_external(Array<T>& v)
+{
+  ASSERT_TRUE(v.isExternal());
+
+  /* Check that the array is full. */
+  ASSERT_EQ(v.size(), v.capacity());
+
+  const IndexType size = v.size();
+  const IndexType num_values = size;
+  T* const data_ptr = v.data();
+
+  /* Set the elements in the array. */
+  for(IndexType i = 0; i < size; ++i)
+  {
+    v[i] = i;
+  }
+
+  /* Check the elements using the raw pointer. */
+  for(IndexType i = 0; i < num_values; ++i)
+  {
+    EXPECT_EQ(data_ptr[i], i);
+  }
+
+  /* Set the elements using the raw pointer. */
+  for(IndexType i = 0; i < size; ++i)
+  {
+    data_ptr[i] = i * i;
+  }
+
+  /* Check the elements using the () operator. */
+  for(IndexType i = 0; i < size; ++i)
+  {
+    EXPECT_EQ(v[i], i * i);
+  }
+
+  EXPECT_EQ(size, v.size());
+  EXPECT_EQ(size, v.capacity());
+  EXPECT_EQ(data_ptr, v.data());
+
+  /* Since the array is full all of the following calls should require a
+   * reallocation and cause a fatal error. */
+  T val = T();
+  EXPECT_DEATH_IF_SUPPORTED(v.push_back(val), IGNORE_OUTPUT);
+  EXPECT_DEATH_IF_SUPPORTED(v.insert(0, 1, val), IGNORE_OUTPUT);
+  EXPECT_DEATH_IF_SUPPORTED(v.reserve(size + 1), IGNORE_OUTPUT);
+}
+
+} /* end namespace internal */
+
+//------------------------------------------------------------------------------
+// UNIT TESTS
+//------------------------------------------------------------------------------
+
+//------------------------------------------------------------------------------
+TEST(core_array, checkStorage)
+{
+  constexpr IndexType ZERO = 0;
+
+  for(IndexType capacity = 2; capacity < 512; capacity *= 2)
+  {
+    Array<int> v_int(ZERO, capacity);
+    internal::check_storage(v_int);
+
+    Array<double> v_double(ZERO, capacity);
+    internal::check_storage(v_double);
+  }
+}
```

Comment:
Suggestion for a future refactor: Perhaps the tests in this file should be set up to use a templated parametrized test fixture, as in https://github.com/LLNL/axom/blob/develop/src/axom/quest/tests/quest_inout_interface.cpp